### ISSUE TITLE
Game Context (Playoffs, Regular Season) and Advanced Player Stats From Individual Games

### ISSUE THREAD

**Issue by benjaminmesser** [jaebradley/basketball_reference_web_scraper#284]
My ultimate goal is to build a database of all significant game data from all previous seasons using this tool. To start, I was planning on scraping from all games using the `team_box_scores` and `player_box_scores` functions to obtain team and player data for every game.

Now one issue is that advanced stats are missing for both of these. Adding them to `team_box_scores` doesn't seem like it would be too challenging, but for `player_box_scores` that data is missing from the "Daily Stats Leaders" pages that it uses, so it would probably have to be reworked to also look at the game box scores like `team_box_scores` does to get advanced player data, which would be a little annoying to implement.

But another issue is getting game context. Seemingly, there is no easy way to find if a game is a regular season or playoff game currently, and ideally, I would want even more context than this, such as which # game it is in the regular season or in a playoff series, and in which playoff series it is. At the very least, it seems like a necessity to know at least if a game is in the regular season game or the playoffs.

Playoff games do have headers like this on them:

![image](https://github.com/user-attachments/assets/73ea2b18-4133-4dd0-b257-b1eaa3fd90f1)

It seems like a pretty janky solution, but is the only way to figure out if a game is a playoff game to scrape these titles?

Or is a better solution to just use some master list of games and dates and contexts from somewhere and then compare all games I have in my database with this list to figure out which games are playoff and which are regular season? This would probably necessitate that the deprecated/defunct teams are figured out/fixed as well because a lot of older games simply have blank team names still. Another issue with this I could see is with teams that played multiple times on the same day, but as far as I can tell, this has only happened once: https://www.statmuse.com/nba/ask/has-a-nba-team-ever-played-2-games-in-a-day so that edge case should be doable to manage. The stats leaders page for that day also seems to work just fine, with players showing up twice if they played in two games that day: https://www.basketball-reference.com/friv/dailyleaders.cgi?month=3&day=8&year=1954

There are also other quirks like all-star games, which are currently missing entirely. Although there is this page so it would probably be fairly simple to add a function to look them up: https://www.basketball-reference.com/allstar/

Either way, do you have any suggestions for any of this? I originally thought I would just build my own web scraper, but after seeing that this project exists and is actively maintained, I'd rather just use this if possible.

**Comment by jaebradley** [comment]
Hey @benjaminmesser thanks for opening up this discussion issue!

I have a couple thoughts (in no particular order)

>all previous seasons using this tool.

Depending on the oldest seasons you want to capture data for, this approach is currently bounded by a flaw in how the scraper handles deprecated teams (it doesn't).

So I need to add support for deprecated teams if you wanted to capture data for seasons all the way back to the very first NBA season.

>but for player_box_scores that data is missing from the "Daily Stats Leaders" pages that it uses, so it would probably have to be reworked to also look at the game box scores like team_box_scores does to get advanced player data, which would be a little annoying to implement.

The issue is in order to get the same view, you'd have to make `X` requests for all `X` games on a given day for each game box score. With this approach, I think you'll quickly run into [the rate limits](https://www.sports-reference.com/bot-traffic.html)

<img width="945" alt="image" src="https://github.com/user-attachments/assets/db7cf84b-cb57-4500-b958-29dfac63498a">

>It seems like a pretty janky solution, but is the only way to figure out if a game is a playoff game to scrape these titles?

Unfortunately, from my cursory view of the site, I also see your suggested approach as the best way to determine if a game is a playoff game or regular season game.

Perhaps using [the unofficial NBA.com stats API](https://github.com/swar/nba_api) may bear more fruit.

**Comment by benjaminmesser** [comment]
> The issue is in order to get the same view, you'd have to make X requests for all X games on a given day for each game box score. With this approach, I think you'll quickly run into the rate limits

Yeah I've already encountered this and just added a delay to avoid getting rate limited. I've already got all of the basic team box scores up to about 2006 after a couple days so it's not too bad. And this pain is just to get all of the historical data, once I'm just doing daily requests to get the new data it should be near-instant. I might need to redo this historical scrape if/when the deprecated team issue is fixed, but it'll have to do for now.

> Perhaps using the unofficial NBA.com stats API may bear more fruit.

Unfortunately, after looking through for a bit and trying some of the API functions, it seems like this is quite limited on game specific data and without any of the advanced stats as well, so idk how useful it would be.

Ultimately, I think I can add functions to get team and player advanced stats to this package. I might try to get the shot data as well because that also looks interesting. I should have time to work on that later this week. But the deprecated teams are another issue that it would be nice to eventually resolve and I'm still not sure what the best approach is for dealing with them. I'll think on it more this week maybe. I'll also keep thinking about game context because I think there has to be some way to get more info on this as well.

**Comment by jaebradley** [comment]
>it seems like this is quite limited on game specific data and without any of the advanced stats as well, so idk how useful it would be.

Interesting, because the available NBA data [endpoints](https://github.com/swar/nba_api/tree/master/docs/nba_api/stats/endpoints)  seem quite extensive.
